Improve Control4 connection error logging

davidrecordon opened the pull request:
## Proposed change
Break apart these connection errors and include debugging context.

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: 
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

davidrecordon committed:
Improve Control4 connection error logging

zweckj reviewed (changes requested):
I believe it would be better to (also) raise the correct exceptions so that it directly shows in the frontend

davidrecordon commented:
@zweckj makes sense, but that's a larger refactoring versus at least getting this detail into logs helped me last night when I was developing another PR and running into a very generic error message. I don't mind doing that other refactor to get more useful error messages back into the configuration UI, but would prefer to split that apart into another PR and land this first for development.

zweckj commented:
Not really? You just need to raise `InvalidAuth`/`CannotConnect` instead of returning `False` it seems.

davidrecordon commented:
Go read the try block starting on line 133, that's raising the exceptions from these two underlying functions so wouldn't we want to refactor across both?

joostlek commented:
It feels weird that whenever you have an unauthenticated error while connecting to the director, you get cannot_connect instead of invalid_auth.

As a sidenote, we should only log things we genuinely don't know. If we receive an unauthenticated issue, then we know something is wrong with the auth, so we don't have to log that as well

davidrecordon committed:
Refactor error messages and logging

davidrecordon committed:
Extract human-readable error messages from pyControl4 responses

davidrecordon committed:
Re-improve code coverage

zweckj commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -46,106 +43,133 @@
 
-class Control4Validator:
-    """Validates that config details can be used to authenticate and communicate with Control4."""
+def _extract_error_message(err: Exception) -> str:
```
I think you went a bit over the top with this one 😅 This complicates it a bit too much, it would’ve just been fine to match the correct UI strings with the correct errors. Because now you are mixing translated and untranslated strings in the UI, which we shouldn’t do or you get something like „Keine Steuereinheit gefunden: something went wrong“ in a German UI which is weird. The translated part should be enough to identify the error and if we don’t know the error source, we can additionally log it

davidrecordon replied:
I think this is where we're running into an issue of a user audience in the UI and a developer audience in the logs. Since the guidance was not to log then the only place to get additional details is back through raising these exceptions in the UI. But we neither know the breadth of what will come back from the API nor then have a way to translate it. This was a better approach than just showing raw JSON in the UI which is what was happening when I was outputting the error itself.

So I think we have generally three options:
1. Remove this from the UI but add some developer-oriented logging back (which sounds more like a best practice anti-pattern these days).
2. Accept that we'll have a translated friendly error paired with an untranslated error from Control4's API.
3. Do neither and have people working on this integration monkey-patch in local logging when they need more details to debug.

I was doing #3 which is what led me to putting up this PR in the first place as I was trying to fix a user reported bug.

davidrecordon committed:
Adjust for known/unknown error handling in Control4 config flow

- Use the common localized string for unknown errors
- Log full stack trace only for unknown exceptions
- Use translatable messages for known errors and exclude raw API details

joostlek committed:
Merge branch 'dev' into control4-connection-logging

joostlek committed:
Fix
